python/GafferUI/SpreadsheetUI/_ProxyModels.py

```diff
@@ -60,166 +69,122 @@ def setSourceModel( self, model ) :
 
-	def rowsPlug( self ) :
-
-		return self.sourceModel().rowsPlug()
-
-	def plugForIndex( self, index ) :
-
-		return self.sourceModel().plugForIndex( self.mapToSource( index ) )
-
-	def valuePlugForIndex( self, index ) :
-
-		return self.sourceModel().valuePlugForIndex( self.mapToSource( index ) )
-
-	def indexForPlug( self, plug ) :
-
-		return self.mapFromSource( self.sourceModel().indexForPlug( plug ) )
-
-	def _dataChanged( self, topLeft, bottomRight, roles ) :
-
-		proxyTopLeft = self.mapFromSource( topLeft )
-		if not proxyTopLeft.isValid() :
-			proxyTopLeft = self.index( 0, 0 )
-
-		proxyBottomRight = self.mapFromSource( bottomRight )
-		if not proxyBottomRight.isValid() :
-			proxyBottomRight = self.index( self.rowCount() - 1, self.columnCount() - 1 )
-
-		self.dataChanged.emit( proxyTopLeft, proxyBottomRight, roles )
-
-class RowNamesProxyModel( __PlugTableProxyModel ) :
-
-	def columnCount( self, parent ) :
+	def columnCount( self, parent = QtCore.QModelIndex() ) :
 
 		if parent.isValid() :
 			return 0
 
-		return 2
+		if self.__columnCount :
+			return self.__columnCount
+		else :
+			return self.sourceModel().columnCount() - self.__startColumn
 
-	def rowCount( self, parent ) :
+	def rowCount( self, parent = QtCore.QModelIndex() ) :
 
 		if parent.isValid() :
 			return 0
 
-		return self.sourceModel().rowCount() - 1
+		if self.__rowCount :
+			return self.__rowCount
+		else :
+			return self.sourceModel().rowCount() - self.__startRow
 
 	def mapFromSource( self, sourceIndex )  :
 
-		if not sourceIndex.isValid() or sourceIndex.row() < 1 :
+		if not sourceIndex.isValid() :
 			return QtCore.QModelIndex()
 
-		if sourceIndex.column() > 1 :
-			return QtCore.QModelIndex()
+		row = sourceIndex.row() - self.__startRow
+		column = sourceIndex.column() - self.__startColumn
 
-		return self.index( sourceIndex.row() - 1, sourceIndex.column() )
-
-	def mapToSource( self, proxyIndex ) :
-
-		if not proxyIndex.isValid() or proxyIndex.row() < 0 :
+		if row < 0 or row >= self.rowCount() :
 			return QtCore.QModelIndex()
 
-		return self.sourceModel().index( proxyIndex.row() + 1, proxyIndex.column() )
-
-	def _dataChanged( self, topLeft, bottomRight, roles ) :
+		if column < 0 or column >= self.columnCount() :
+			return QtCore.QModelIndex()
 
-		# Any region after the enabled column can never affect our model
-		if topLeft.column() > 1 :
-			return
+		return self.index( row, column )
 
-		# Anything in the default row can never affect our model
-		if topLeft.row() == 0 and bottomRight.row() == 0 :
-			return
+	def mapToSource( self, proxyIndex ) :
 
-		super( RowNamesProxyModel, self )._dataChanged( topLeft, bottomRight, roles )
+		if not proxyIndex.isValid():
+			return QtCore.QModelIndex()
 
-class DefaultsProxyModel( __PlugTableProxyModel ) :
+		row = proxyIndex.row() + self.__startRow
+		column = proxyIndex.column() + self.__startColumn
 
-	def columnCount( self, parent ) :
+		if row < 0 or row >= self.sourceModel().rowCount() :
```

Comment:
How can `row < 0` be true? It should have started positive, and can only have had positive numbers added to it. Also, if the proxy index is valid, how can `row >= sourceModel.rowCount()`?

Reply:
Good spot, thanks [removed](https://github.com/GafferHQ/gaffer/pull/3995/commits/fb03391e3f83ca0f88189e20086909e3b0c3e520).

Reply:
How about `row >= sourceModel.rowCount()`? Am I missing some reason that could be true for a valid proxy index?